File: XCTestBootstrapTests/Fixtures/FBXCTestReporterDouble.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBXCTestReporterDouble.h"

@interface FBXCTestReporterDouble ()

@property (nonatomic, strong, readonly) NSMutableArray<NSString *> *mutableStartedSuites;
@property (nonatomic, strong, readonly) NSMutableArray<NSString *> *mutableEndedSuites;
@property (nonatomic, strong, readonly) NSMutableArray<NSArray<NSString *> *> *mutableStartedTestCases;
@property (nonatomic, strong, readonly) NSMutableArray<NSArray<NSString *> *> *mutablePassedTests;
@property (nonatomic, strong, readonly) NSMutableArray<NSArray<NSString *> *> *mutableFailedTests;
@property (nonatomic, strong, readonly) NSMutableArray<NSDictionary *> *mutableExternalEvents;
@property (nonatomic, assign, readwrite) BOOL printReportWasCalled;

@end

@implementation FBXCTestReporterDouble

#pragma mark Initializers

- (instancetype)init
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _mutableStartedSuites = [NSMutableArray array];
  _mutableEndedSuites = [NSMutableArray array];
  _mutableStartedTestCases = [NSMutableArray array];
  _mutablePassedTests = [NSMutableArray array];
  _mutableFailedTests = [NSMutableArray array];
  _mutableExternalEvents = [NSMutableArray array];

  return self;
}

#pragma mark Methods for Validating

- (void)testCaseDidStartForTestClass:(NSString *)testClass method:(NSString *)method
{
  [self.mutableStartedTestCases addObject:@[testClass, method]];
}

- (BOOL)printReportWithError:(NSError **)error
{
  self.printReportWasCalled = YES;
  return YES;
}

- (void)handleExternalEvent:(NSString *)line
{
  NSError *error = nil;
  NSDictionary *event = [NSJSONSerialization JSONObjectWithData:[line dataUsingEncoding:NSUTF8StringEncoding] options:0 error:&error];
  if (event == nil) {
    return;
  }
  [self.mutableExternalEvents addObject:event];
}

#pragma mark Stubbed Methods

- (void)didBeginExecutingTestPlan
{

}

- (void)testSuite:(NSString *)testSuite didStartAt:(NSString *)startTime
{
  [self.mutableStartedSuites addObject:testSuite];
}

- (void)finishedWithSummary:(FBTestManagerResultSummary *)summary
{
  [self.mutableEndedSuites addObject:summary.testSuite];
}

- (void)didFinishExecutingTestPlan
{

}

- (void)testHadOutput:(NSString *)output
{

}

- (void)processWaitingForDebuggerWithProcessIdentifier:(pid_t)pid
{

}

- (void)didRecordVideoAtPath:(nonnull NSString *)videoRecordingPath
{

}

- (void)didSaveOSLogAtPath:(nonnull NSString *)osLogPath
{

}

- (void)didCopiedTestArtifact:(nonnull NSString *)testArtifactFilename toPath:(nonnull NSString *)path
{

}

- (void)processUnderTestDidExit
{

}

- (void)testCaseDidFinishForTestClass:(nonnull NSString *)testClass method:(nonnull NSString *)method withStatus:(FBTestReportStatus)status duration:(NSTimeInterval)duration logs:(nullable NSArray<NSString *> *)logs
{
  NSArray<NSString *> *pairs = @[testClass, method];

  switch (status) {
    case FBTestReportStatusPassed:
      [self.mutablePassedTests addObject:pairs];
      return;
    case FBTestReportStatusFailed:
      [self.mutableFailedTests addObject:pairs];
      return;
    default:
      return;
  }
}

- (void)didCrashDuringTest:(NSError *)error
{
  
}

- (void)testCaseDidFailForTestClass:(nonnull NSString *)testClass method:(nonnull NSString *)method exceptions:(nonnull NSArray<FBExceptionInfo *> *)exceptions {
  
}


#pragma mark Accessors

- (NSArray<NSString *> *)startedSuites
{
  return [self.mutableStartedSuites copy];
}

- (NSArray<NSString *> *)endedSuites
{
  return [self.mutableEndedSuites copy];
}

- (NSArray<NSArray<NSString *> *> *)startedTests
{
  return [self.mutableStartedTestCases copy];
}

- (NSArray<NSArray<NSString *> *> *)passedTests
{
  return [self.mutablePassedTests copy];
}

- (NSArray<NSArray<NSString *> *> *)failedTests
{
  return [self.mutableFailedTests copy];
}

- (NSArray<NSDictionary *> *)eventsWithName:(NSString *)name
{
  NSPredicate *predicate = [NSPredicate predicateWithBlock:^ BOOL (NSDictionary<NSString *, id> *event, id _) {
    return [event[@"event"] isEqualToString:name];
  }];
  return [self.mutableExternalEvents filteredArrayUsingPredicate:predicate];
}

@end

File: XCTestBootstrapTests/Fixtures/FBXCTestReporterDouble.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import XCTestBootstrap

final class FBXCTestReporterDouble: NSObject, FBXCTestReporter {

  private var mutableStartedSuites: [String] = []
  private var mutableEndedSuites: [String] = []
  private var mutableStartedTestCases: [[String]] = []
  private var mutablePassedTests: [[String]] = []
  private var mutableFailedTests: [[String]] = []
  private var mutableExternalEvents: [[String: Any]] = []
  private(set) var printReportWasCalled = false

  var logDirectoryPath: String?

  // MARK: - Accessors

  var startedSuites: [String] {
    return mutableStartedSuites
  }

  var endedSuites: [String] {
    return mutableEndedSuites
  }

  var startedTests: [[String]] {
    return mutableStartedTestCases
  }

  var passedTests: [[String]] {
    return mutablePassedTests
  }

  var failedTests: [[String]] {
    return mutableFailedTests
  }

  func events(withName name: String) -> [[String: Any]] {
    return mutableExternalEvents.filter { event in
      (event["event"] as? String) == name
    }
  }

  // MARK: - FBXCTestReporter

  func testCaseDidStart(forTestClass testClass: String, method: String) {
    mutableStartedTestCases.append([testClass, method])
  }

  func printReport() throws {
    printReportWasCalled = true
  }

  func handleExternalEvent(_ line: String) {
    guard let data = line.data(using: .utf8),
      let event = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else {
      return
    }
    mutableExternalEvents.append(event)
  }

  func didBeginExecutingTestPlan() {}

  func testSuite(_ testSuite: String, didStartAt startTime: String) {
    mutableStartedSuites.append(testSuite)
  }

  func finished(with summary: FBTestManagerResultSummary) {
    mutableEndedSuites.append(summary.testSuite)
  }

  func didFinishExecutingTestPlan() {}

  func testHadOutput(_ output: String) {}

  func processWaitingForDebugger(withProcessIdentifier pid: pid_t) {}

  func didRecordVideo(atPath videoRecordingPath: String) {}

  func didSaveOSLog(atPath osLogPath: String) {}

  func didCopiedTestArtifact(_ testArtifactFilename: String, toPath path: String) {}

  func processUnderTestDidExit() {}

  func testCaseDidFinish(forTestClass testClass: String, method: String, with status: FBTestReportStatus, duration: TimeInterval, logs: [String]?) {
    let pairs = [testClass, method]
    switch status {
    case .passed:
      mutablePassedTests.append(pairs)
    case .failed:
      mutableFailedTests.append(pairs)
    case .unknown:
      break
    @unknown default:
      break
    }
  }

  func didCrashDuringTest(_ error: Error) {}

  func testCaseDidFail(forTestClass testClass: String, method: String, exceptions: [FBExceptionInfo]) {}
}
